desktop/createDownloadQueue.ts

```diff
@@ -16,21 +16,46 @@ type DownloadItem = {
 };
 
 const createDownloadQueue = () => {
-    const downloadItem = (item: DownloadItem): Promise<void> =>
-        new Promise((resolve) => {
-            const options = {
-                ...item.options,
-                onStarted: () => {
-                    item.win.webContents.send(ELECTRON_EVENTS.DOWNLOAD_STARTED, {url: item.url});
-                },
-                onCompleted: () => resolve(),
-                onCancel: () => resolve(),
+    const downloadItemProcessor = (item: DownloadItem): Promise<void> =>
+        new Promise((resolve, reject) => {
+            item.win.webContents.downloadURL(item.url);
+
+            const listener = (event: Electron.Event, downloadItem: Electron.DownloadItem) => {
+                const cleanup = () => item.win.webContents.session.removeListener('will-download', listener);
+                const errorMessage = `The download of ${downloadItem.getFilename()} was interrupted`;
+
+                downloadItem.on('updated', (_, state) => {
+                    if (state !== 'interrupted') {
+                        return;
+                    }
+
+                    cleanup();
+                    reject(new Error(errorMessage));
+                    downloadItem.cancel();
+                });
+
+                downloadItem.on('done', (_, state) => {
+                    cleanup();
+                    if (state === 'cancelled') {
+                        resolve();
+                    } else if (state === 'interrupted') {
+                        reject(new Error(errorMessage));
+                    } else if (state === 'completed') {
+                        if (process.platform === 'darwin') {
+                            const savePath = downloadItem.getSavePath();
+                            app.dock.downloadFinished(savePath);
+                        }
+                        resolve();
+                    }
+                });
+
+                item.win.webContents.send(ELECTRON_EVENTS.DOWNLOAD_STARTED, {url: item.url});
             };
 
-            electronDownload(item.win, item.url, options);
+            item.win.webContents.session.on('will-download', listener);
```

Comment:
Can we not save the listener in a variable? Would be a little more readable.

Reply:
No, use it directly like:
```
            item.win.webContents.session.on('will-download', () => {
                // code for the listener
            });
```

Reply:
how can we perform cleanup without the `listener` variable?

Reply:
Ah, forgot about that. Never mind then.

Reply:
Please do not approve yet. I have noticed that the file has been renamed to download.html, and I am investigating it.

Reply:
> Please do not approve yet. I have noticed that the file has been renamed to download.html, and I am investigating it.

@allroundexperts ok it's fixed, please help review the recent changes